src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -5824,15 +5824,15 @@ private bool IsPossibleName()
             return this.IsTrueIdentifier();
         }
 
-        private ScanTypeFlags ScanType()
+        private ScanTypeFlags ScanType(bool forPattern = false)
```

Comment:
@gafter I have a suggested refactoring to make (though it is almost certainly not an appropriate time to perform that suggestion).  Specifically, I believe the existing system for ScanXXX+ParseXXX in LanguageParser amounts to a huge amount of complexity for little (if any) gain.  I have an alternate approach i think would be superior and better for the long term health of the parser.

The primary problem today is that the parser needs to encode all the parsing rules all over again into the ScanXXX functions , keeping both in sync, and trying to ensure that the appropriate results from scanning carry over to parsing. 

There is an alternative approach that can be taken, which we followed in TypeScript.  You can see that approach here:
https://github.com/Microsoft/TypeScript/blob/master/src/compiler/parser.ts#L1020  (example explained in more depth below).

Specifically, instead of actually having a separate scanning step, we just went ahead and tried the parse, passing in whatever flags were appropriate for the current parse context, and specifying what behavior we wanted if the parse ran into issues.  Effectively, with LanguageParser, this would just be using reset points, and asking to rollback once enough information was gleaned from the parse to determine that the wrong speculative path had been taken.

The concern with this approach was of course one of allocation cost.  i.e. would the cost of allocations be a hindrance to parsing speed.  In practice, we found that cost to be completely negligible in most cases, and easily worked around in the rare cases where it did impact.  Specifically, we found that the majority of code tends to follow a more common case, and that case should be what the speculative parsing code looks for first.  In the rare cases where that doesn't follow, usually a little lookahead (which is vastly less complex than the arbitrary 'scanning' subsystem that we have) can allow for a sensible direction to take.  And, when that doesn't suffice, you do just do the arbitrary parsing+rollback.

--

I only bring this up because i know how tricky it's been over time to try to get scanning+parsing to work with these more and more complex expression constructs.  While it would not be trivial to move off of scanning, I do think it would make for a vastly simpler system, and would actually pay dividends later as new language features would be easier to incorporate.

Anyways, just my 2c.  If you'd like to discuss this further (including offline, or not in this PR) I'm happy to assist.

--

Example: explained.  

Here's an example of how TS deals with the ```<``` ambiguity we also have in C#:

```c#
                    // See if this is the start of a generic invocation.  If so, consume it and
                    // keep checking for postfix expressions.  Otherwise, it's just a '<' that's
                    // part of an arithmetic expression.  Break out so we consume it higher in the
                    // stack.
                    const typeArguments = tryParse(parseTypeArgumentsInExpression);
...
        function parseTypeArgumentsInExpression() {
            if (!parseOptional(SyntaxKind.LessThanToken)) {
                return undefined;
            }

            const typeArguments = parseDelimitedList(ParsingContext.TypeArguments, parseType);
            if (!parseExpected(SyntaxKind.GreaterThanToken)) {
                // If it doesn't have the closing >  then it's definitely not an type argument list.
                return undefined;
            }

            // If we have a '<', then only parse this as a argument list if the type arguments
            // are complete and we have an open paren.  if we don't, rewind and return nothing.
            return typeArguments && canFollowTypeArgumentsInExpression()
                ? typeArguments
                : undefined;
        }
```

Here, there's no need for a separate ScanTypeArgumentList function.  TypeScript simply goes through and actually attempts the parse, easily falling back if it fails.  Because the logic is just in a single place, there aren't two codepaths to keep in sync.  Also, this code can be extremely flexible.  For example, we could have code like the above, but extend it so that if we see things like primitive types (i.e. int/string/etc.) in the delimited list, then we would always consider it a type argument list.  i.e. ```List<int>``` would always parse as a type-arg list, even though it isn't followed by something like ```(```.

Right now we accomplish that in the scanning phase, but it's extremely tricky to get right.  By moving it right into the parsing phase, we can do much richer analysis as we have the full tree up to that point (as well as any future tokens) to look at at make determinations.

--

In the end, everything can be done with ScanXXX.  But it really does add lots of complexity (esp. as expressions have gotten a lot more ambiguous).  Moving away from this into a simpler "just do the parse and make the decisions based on how it's gone and what you can see coming up" can make life much simpler and easier.

Reply:
@CyrusNajmabadi The parsing code here closely models the way expressions are parsed. The reason that we have this in the compiler is due to a commonly occurring ambiguity in the language that occurs when we encounter an open paren (which starts different contexts like a cast or a parenthesized subexpression), and which occurs very frequently. With pattern matching, declaration expressions, and deconstruction the ambiguity occurs even more often. We would have to measure the performance (and memory) impact of parsing eagerly instead of scanning before we could move to that model. In the current code base the trial parse *usually* fails, so it would definitely come with a runtime cost. That experiment should be done separately rather than as part of this PR.

Reply:
> That experiment should be done separately rather than as part of this PR.

100% agree.  I just wanted to share some ideas, but didn't know the best way to do so other than suggesting them in this PR.  I def would not expect anything to come of this for quite some time.  It was just something to keep in the back of your head if you ever start feeling that the current complexity is becoming onerous.

Thanks!

Reply:
> commonly occurring ambiguity in the language that occurs when we encounter an open paren (which starts different contexts like a cast or a parenthesized subexpression), and which occurs very frequently

Indeed.  That would be one of those places that i talked about where we might have a simple lookahead system solely for determining which direction in the ambiguity to prefer; but always knowing we might be wrong and might need to fallback to the other path.

But, again, i totally agree.  This is not the right time for such a change.  And it may never be valuable enough to make this change.  Thanks for listening :)